src/controller_examples/rabbitmq_controller/proof/liveness/helper_invariants/owner_ref.rs

```diff
@@ -0,0 +1,297 @@
+// Copyright 2022 VMware, Inc.
+// SPDX-License-Identifier: MIT
+#![allow(unused_imports)]
+use super::invariants::*;
+use crate::external_api::spec::*;
+use crate::kubernetes_api_objects::{
+    api_method::*, common::*, config_map::*, dynamic::*, owner_reference::*, resource::*,
+    stateful_set::*,
+};
+use crate::kubernetes_cluster::spec::{
+    builtin_controllers::types::{built_in_controller_req_msg, BuiltinControllerChoice},
+    cluster::*,
+    cluster_state_machine::Step,
+    controller::common::{controller_req_msg, ControllerActionInput, ControllerStep},
+    message::*,
+};
+use crate::rabbitmq_controller::{
+    common::*,
+    proof::common::*,
+    spec::{rabbitmqcluster::*, reconciler::*},
+};
+use crate::temporal_logic::{defs::*, rules::*};
+use vstd::prelude::*;
+
+verus! {
+
+spec fn sts_key_not_exists_or_delete_msg_in_flight_or_sts_has_current_ref(rabbitmq: RabbitmqClusterView) -> StatePred<RMQCluster> {
+    let sts_key = make_stateful_set_key(rabbitmq.object_ref());
+    |s: RMQCluster| {
+        ||| {
+            &&& s.resource_key_exists(sts_key)
+            &&& (
+                    (exists |msg: Message| {
+                        #[trigger] s.message_in_flight(msg)
+                        && msg.dst.is_KubernetesAPI()
+                        && msg.content.is_delete_request_with_key(sts_key)
+                    })
+                    || s.resource_obj_of(sts_key).metadata.owner_references == Some(seq![rabbitmq.controller_owner_ref()])
+            )
+        }
+        ||| !s.resource_key_exists(sts_key)
+    }
+}
+
+#[verifier(external_body)]
+pub proof fn lemma_eventually_only_valid_server_config_map_exists(spec: TempPred<RMQCluster>, rabbitmq: RabbitmqClusterView)
+    requires
+        rabbitmq.well_formed(),
+        spec.entails(always(lift_state(RMQCluster::busy_disabled()))),
+        spec.entails(always(lift_action(RMQCluster::next()))),
+        spec.entails(tla_forall(|i| RMQCluster::kubernetes_api_next().weak_fairness(i))),
+        spec.entails(tla_forall(|i| RMQCluster::builtin_controllers_next().weak_fairness(i))),
+        spec.entails(always(lift_state(RMQCluster::desired_state_is(rabbitmq)))),
+        spec.entails(always(lift_state(server_config_map_has_no_finalizers_or_timestamp_and_only_has_controller_owner_ref(rabbitmq)))),
+        spec.entails(always(lift_state(every_update_cm_req_does_the_same(rabbitmq)))),
+        spec.entails(always(lift_state(every_create_cm_req_does_the_same(rabbitmq)))),
+    ensures
+        spec.entails(true_pred().leads_to(always(lift_state(server_config_map_has_owner_reference_pointing_to_current_cr(rabbitmq))))),
+{}
+
+pub proof fn lemma_eventually_only_valid_stateful_set_exists(spec: TempPred<RMQCluster>, rabbitmq: RabbitmqClusterView)
```

Comment:
Could you write some comments for this lemma, since this is the first time we reason about gc? This lemma would be a very valuable reference for later proof.

Reply:
I want to refactor this lemma to make it more general and move it to some common module. But I can add some temporary comments.

Reply:
Looking forward